semantic-python/src/Language/Python/Failure.hs

```diff
@@ -0,0 +1,53 @@
+{-# LANGUAGE DeriveFoldable, DeriveFunctor, DeriveTraversable, ExistentialQuantification, FlexibleContexts,
+             KindSignatures, LambdaCase, MultiParamTypeClasses, QuantifiedConstraints, RankNTypes, StandaloneDeriving,
+             TypeOperators #-}
+
+module Language.Python.Failure
+  ( Failure (..)
+  , unimplemented
+  , invariantViolated
+  , eliminateFailures
+  ) where
+
+import Prelude hiding (fail)
+
+import Control.Effect.Carrier
+import Control.Monad.Fail
+import Data.Coerce
+import Data.Kind
+import Syntax.Module
+import Syntax.Term
+import Syntax.Traversable
+
+data Failure (f :: Type -> Type) a
+  = forall a . Show a => Unimplemented a
```

Comment:
I have been hemming and hawing over whether we should use a `String` instead of the existential:

1. Existentials are a bit of a pain to work with in terms of deriving &c.

2. The `Show` constraint means that literally the only thing we can do with the value is show it—tho unlike a `String`, we _can_ show it at different precedence levels, if we care about that.

3. Laziness means that if we don’t ask for the `String` then we won’t have done any work.

4. `String` will allocate a thunk. But the existential stores a dict. We don’t expect this to be the common case anyway, so this isn’t a super compelling argument in either direction IMO.

5. If we wanted to be able to do other things with the value, then the existential makes that easy, but… YAGNI and all that.

Ultimately I think we’re better off with the `String`, for now. Thoughts?

Reply:
Yeah, I don’t object to the `String`. I was just flexing because I could.

Reply:
Because there’s an effectful interface in front of this, we can change the effect type without bothering the callosities.